src/language-html/print/block.js

```diff
@@ -0,0 +1,123 @@
+import { group, hardline, indent } from "../../document/builders.js";
+import { printChildren } from "./children.js";
+
+let uid = 0;
+
+const settings = {
+  if: {
+    withExpression: true,
+    isFollowingBlock: false,
+    followingBlocks: ["else if", "else"],
+  },
+  "else if": {
+    withExpression: true,
+    isFollowingBlock: true,
+    followingBlocks: ["else if", "else"],
+  },
+  else: {
+    withExpression: false,
+    isFollowingBlock: true,
+    followingBlocks: [],
+  },
+
+  switch: {
+    withExpression: true,
+    isFollowingBlock: false,
+    followingBlocks: [],
+  },
+  case: {
+    withExpression: true,
+    isFollowingBlock: false,
+    followingBlocks: [],
+  },
+  default: {
+    withExpression: false,
+    isFollowingBlock: false,
+    followingBlocks: [],
+  },
+
+  for: {
+    withExpression: true,
+    isFollowingBlock: false,
+    followingBlocks: ["empty"],
+  },
+  empty: {
+    withExpression: false,
+    isFollowingBlock: true,
+    followingBlocks: [],
+  },
+
+  defer: {
+    withExpression: true,
+    isFollowingBlock: false,
+    followingBlocks: ["placeholder", "error", "loading"],
+  },
+  placeholder: {
+    withExpression: false,
+    isFollowingBlock: true,
+    followingBlocks: ["placeholder", "error", "loading"],
+  },
+  error: {
+    withExpression: false,
+    isFollowingBlock: true,
+    followingBlocks: ["placeholder", "error", "loading"],
+  },
+  loading: {
+    withExpression: true,
+    isFollowingBlock: true,
+    followingBlocks: ["placeholder", "error", "loading"],
+  },
+};
+
+function printBlock(path, options, print) {
+  const { node } = path;
+
+  const setting = settings[node.name];
+  const contents = [];
+
+  if (!setting) {
+    throw new Error("Unknown block name: " + node.name);
+  }
+
+  if (
+    setting.isFollowingBlock &&
+    path.previous &&
+    path.previous.type === "block"
+  ) {
+    contents.push("} ");
+  }
+
+  contents.push(`@${node.name}`);
+
+  if (setting.withExpression) {
+    const expression = fetchExpression(node);
+    contents.push(` (${expression})`);
+  }
+
+  contents.push(" {");
+
+  const children = printChildren(path, options, print);
+  contents.push(indent([hardline, children]));
+
+  if (shouldCloseBlock(node, setting.followingBlocks)) {
+    contents.push(hardline, "}");
+  }
+
+  return group(contents, {
+    id: Symbol("block-" + ++uid),
+    shouldBreak: true,
+  });
+}
+
+export { printBlock };
+
+function shouldCloseBlock(node, names) {
+  return names.length === 0 || !node.next || !names.includes(node.next.name);
+}
+
+function fetchExpression(node) {
+  return node.parameters
+    .filter((parameter) => parameter.type === "blockParameter")
+    .map((parameter) => parameter.expression)
```

Comment:
Are the expressions js code? I think the should be formatted as well.

**Prettier pr-15606**
[Playground link](https://deploy-preview-15606--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEABAlgMwAQAp0DOAwgBZxgDWcAJrgDpR0yEB0D22ArgOYA2AngwCUQ7MAYAeAgAcAhlAB8pclWoSA9DPkKGAXwao4vAnDGStigHIQY2MGUo0NFnVF0gANCAjTm0AsigsgBOwRAA7gAKIQgBKLK84bL8AV4ARsGyjjAAyrIAtnAAMuhQcMiYCSbpmdk5cmCl3MgwwZxwXib56C1tHSAETbxwAIqcNuVIlcb9AFYEAB45Q6Pj8BVV-QCOa3CRYdJxILIEALRlNDSeIK2y6LxNRBD5+bLIx7y814NQfHAAgjBWug0px4JE4MESmUNjMvCQYPleAB1EjoeBaMBwHKxdHoABu6P47zABFSIHx7QAklBqAhcmBguhfP9aTkYPxhrDqiBpGETMjMtJ3ny4CZgvjyl5SuKYPtZNxXtz+nJguL3vJuJxeCFrnzSjBkehqDASMgABwABi8wTgO3QtvlireU02XhgsjSRpNZqQACYvJwTAAVT1xaY8uD5NI0OnUIqazgKuAAMQgwVeQKaGrBEBAul0QA)
<!-- prettier-ignore -->
```sh
--parser angular
```

**Input:**
<!-- prettier-ignore -->
```html
@if (isChecked(
	is.
  ugly
)) {
<span>Checked</span>
}
@else {
<span>Not checked</span>
}
```

**Output:**
<!-- prettier-ignore -->
```html
@if (isChecked(
	is.
  ugly
)) {
  <span>Checked</span>
} @else {
  <span>Not checked</span>
}

```

Reply:
Not JS code. Following is a maybe valid Angular code:

```html
@defer (when isVisible() && foo; on hover(button), timer(10s), idle, immediate, interaction(button), viewport(container); prefetch on immediate; prefetch when isDataLoaded()) {
  <calendar-cmp [date]="current" />
} @loading (minimum 1s; after 100ms) {
  Loading...
} @placeholder {
  Placeholder content!
} @error {
  Loading failed :(
}
```

So we need to introduce new embedded formatting.

Reply:
Sorry I don't know

Reply:
`@angular/compiler` parses block parameters here: https://github.com/angular/angular/blob/9cac4eea5b692dd55e43ebcf8237d78777b0deb0/packages/compiler/src/render3/r3_control_flow.ts#L376-L396

Reply:
Let's defer this? Just basic support in this PR.

Reply:
:+1:

Reply:
The functionality entry seem to be  `htmlAstToRender3Ast` https://github.com/angular/angular/blob/9cac4eea5b692dd55e43ebcf8237d78777b0deb0/packages/compiler/src/render3/r3_template_transform.ts#L65

Reply:
> But @if only accept JS code?

It is what Angular call microsyntax. Heavily inspired by JS, but not quite only JS. It was described in https://github.com/angular/angular/blob/3f9033626662130cc666c015793d128d51124f6d/aio/content/guide/structural-directives.md#microsyntax. But it has been removed from documentation in https://github.com/angular/angular/pull/40015, almost 3 years ago. So it might not be a proper source of truth anymore...

Reply:
Using parseTemplateBindings will cause exceptions in `@for` and `@defer`, although my current version uses it (__ng_directive)

Let me study angular-estree-parser, these exceptions may be solved using parseBlockParameterToBinding

Reply:
These codes are about render3 of @angular/compiler,
but currently the angular-html-parser does not contain render3.
I will try to move it.

Reply:
I don't think it is necessary to support the block parameters format in this PR. I think PR is becoming too complex.